### PR TITLE
More magic

### PR DESCRIPTION
**Pull request by btel**
:mage_woman: Magic is all around so lets add more magic! :mage: 

For the list of supported magics see:

https://docs.google.com/spreadsheets/d/1xcBlx8yrNJIrgSQaAbiphr6dNlwd6Dw3a5bzi-Qa9yY/edit?usp=sharing

See also the notebook with examples:

https://github.com/btel/xeus-python/blob/more-magic/notebooks/xeus-magics.ipynb

# Screenshots

## History

peek-history

## Recall

peek-recall

## Timeit
Peek-timeit

## Quick help

Peek 2020-05-28 23-54

### CODE REVIEW

File: src/xcomm.cpp [hunk 1]
```diff
@@ -187,6 +188,44 @@ namespace xpyt
         py::module kernel_module = py::module("kernel");
 
         py::class_<detail::xmock_object> _Mock(kernel_module, "_Mock");
+        py::class_<XInteractiveShell> XInteractiveShell(
```
Comment (JohanMabille) [inline src/xcomm.cpp:191]: Naming convention:

- we use snake case in C++ (even for classes), so this line should probably changed into `py::class_<xinteractive_shell> interactive_shell(// same as before)

- names starting with `_` or `__` are reserved for use by the standard library.

File: src/xinteractiveshell.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            "sx"_a=cell_magics["sx"],
+            "system"_a=cell_magics["system"]);
+
+        magics_manager.attr("magics") = py::dict(
+           "line"_a=line_magics,
+           "cell"_a=cell_magics);
+    }
+
+
+    XInteractiveShell::XInteractiveShell() {
```
Comment (JohanMabille) [inline src/xinteractiveshell.cpp:61]: Formatting: opening curly brace should be on a new line.

File: src/xinteractiveshell.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace xpyt 
+{
+
+    struct hooks_object
+    {
+        hooks_object() {}
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:16]: This can ben removed, the compiler will generate it automatically.

File: src/xinteractiveshell.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        static inline void show_in_pager(py::str data, py::kwargs)
+        {
+            xdisplay(py::dict("text/plain"_a=data), {}, {}, py::dict(), py::none(), py::none(), false, true);
+        }
+
+    };
+
+    class XInteractiveShell {
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:24]: Formatting: opening curly brace should be on a new line

File: src/xinteractiveshell.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    private:
+        py::module _ipy_process;
+        py::module _magic_core;
+        py::module _magics_module;
+        py::module _extension_module;
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:30]: Naming convention: we usually use `m_xxx` for attributes names.

The private section should come after the public one (most of the time you want to read the API, not the implementation details)

File: src/xinteractiveshell.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        py::str ipython_dir;
+
+        //required by pushd
+        py::list dir_stack;
+        py::str home_dir;
+
+        py::object magics_manager;
+        py::object extension_manager;
+
+        hooks_object hooks;
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:50]: Do we need write access for all of these properties? If not, they should be moved to the private section and getters should be provided. You can bind then the getters with `def_property_readonly`.

File: src/xinteractiveshell.hpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        void register_post_execute(py::args, py::kwargs) {};
+        void enable_gui(py::args, py::kwargs) {};
+        void observe(py::args, py::kwargs) {};
+        void showtraceback(py::args, py::kwargs) {};
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:55]: I guess these are required by some python code?

File: src/xinteractiveshell.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        void init_magics();
```
Comment (JohanMabille) [inline src/xinteractiveshell.hpp:57]: This should be moved to the private section

File: src/xcomm.cpp (hunk 1 above)
Comment (btel) [reply src/xcomm.cpp:191]: :+1:  but what do you do when you create a Python class wrapper for a C++ class? I guess you would use CamelCase for Python (like in the line above)?

File: src/xinteractiveshell.hpp (hunk 7 above)
Comment (btel) [reply src/xinteractiveshell.hpp:55]: yes ;-)

**Comment by btel** [comment]
@martinRenou when I change `p_history_manger` to `xeus:xhistory_manager &` and use `p_history_manager=history_manager` in the constructor, I get this error:

```
/home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.cpp:68:29: error: use of deleted function ‘xeus::xhistory_manager& xeus::xhistory_manager::operator=(const xeus::xhistory_manager&)’
   68 |         p_history_manager = history_manager;
      |                             ^~~~~~~~~~~~~~~
In file included from /home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.hpp:4,
                 from /home/bartosz/repos/libraries/xeus-python/src/xinteractiveshell.cpp:1:
/home/bartosz/.pyenv/versions/miniconda3-latest/envs/xeus/include/xeus/xhistory_manager.hpp:32:27: note: declared here
   32 |         xhistory_manager& operator=(const xhistory_manager&) = delete;
```

it seems that someone took care to delete the = operator for references.

**Comment by JohanMabille** [comment]
@btel could you rebase instead of merging to avoid polluting history? Also squashing commits that are not meaningful with the meaningful ones?

**Comment by btel** [comment]
Hi @JohanMabille, sorry I naively thought that it was what the github would do when I click "resolve conflicts" ;-). I squashed the commits into one and rebased them with master.

**Comment by JohanMabille** [comment]
No worries ;) Thanks for rebasing, I'm gonna review it asap!

**Comment by JohanMabille** [comment]
The issue on Windows is unrelated to thie PR. See https://bugs.python.org/issue36020 and https://github.com/nlohmann/json/issues/1408.

**Comment by JohanMabille** [comment]
Thanks a lot!

**Comment by btel** [comment]
Thanks, @JohanMabille . It was lighting fast! :heart_eyes:

**Comment by JohanMabille** [comment]
;) I forgot to ask, why do we need the `py::dynamic_attr()` argument in the interactive shell binding?

**Comment by btel** [comment]
hmm, maybe we don't need it anymore. I used that for testing when I was adding attributes on runtime, but I think now all of them are statically defined.

**Pull request merged**